File: apps/web/lib/target-group-lifecycle.ts
/** Target group lifecycle values (see `apps/api/app/services/target_group_lifecycle.py`). */
export const TARGET_GROUP_LIFECYCLE = {
  active: "active",
  archived: "archived",
} as const;

export type TargetGroupLifecycleStatus =
  (typeof TARGET_GROUP_LIFECYCLE)[keyof typeof TARGET_GROUP_LIFECYCLE];

const LEGACY_ACTIVE = new Set(["draft", "published", "active", ""]);

/** Read status from API/DB — maps legacy draft/published to active. */
export function coerceTargetGroupStatus(
  value?: string | null
): TargetGroupLifecycleStatus {
  const s = (value ?? "").trim().toLowerCase();
  if (s === TARGET_GROUP_LIFECYCLE.archived) {
    return TARGET_GROUP_LIFECYCLE.archived;
  }
  return TARGET_GROUP_LIFECYCLE.active;
}

export function isTargetGroupArchived(value?: string | null): boolean {
  return coerceTargetGroupStatus(value) === TARGET_GROUP_LIFECYCLE.archived;
}

export function isLegacyActiveTargetGroupStatus(value?: string | null): boolean {
  return LEGACY_ACTIVE.has((value ?? "").trim().toLowerCase());
}

File: apps/api/app/services/target_group_lifecycle.py
"""Target group lifecycle — active (default) and archived (hidden from default lists)."""

from __future__ import annotations

_TARGET_GROUP_STATUSES = frozenset({"active", "archived"})
_LEGACY_ACTIVE_STATUSES = frozenset({"draft", "published", "active", ""})


def _norm(value: object | None) -> str:
    if value is None:
        return ""
    if hasattr(value, "value"):
        return str(getattr(value, "value", "") or "").strip().lower()
    return str(value).strip().lower()


def coerce_target_group_status(value: object | None) -> str:
    """Map DB/API values to ``active`` or ``archived`` (legacy draft/published → active)."""

    s = _norm(value)
    if s == "archived":
        return "archived"
    return "active"


def normalize_target_group_status(value: object | None) -> str:
    """Validate writes — only ``active`` or ``archived`` allowed."""

    if value is None or (isinstance(value, str) and not value.strip()):
        return "active"
    s = _norm(value)
    if s in _LEGACY_ACTIVE_STATUSES:
        return "active"
    if s == "archived":
        return "archived"
    raise ValueError("target_group_status_invalid: status must be 'active' or 'archived'")


def is_target_group_archived(value: object | None) -> bool:
    return coerce_target_group_status(value) == "archived"
